driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -308,139 +309,66 @@ public void sendAndReceiveAsync(final CommandMessage message, final SingleResult
+                                             final CommandEventSender commandEventSender) {
+        sendMessageAsync(bsonOutput.getByteBuffers(), messageId, new SingleResultCallback<Void>() {
             @Override
             public void onResult(final Void result, final Throwable t) {
                 bsonOutput.close();
                 if (t != null) {
-                    sendFailedEvent(startTimeNanos, message, commandName, t);
+                    commandEventSender.sendFailedEvent(t);
                     callback.onResult(null, t);
                 } else {
-                    readAsync(MessageHeader.MESSAGE_HEADER_LENGTH, new SingleResultCallback<ByteBuf>() {
+                    readAsync(MESSAGE_HEADER_LENGTH, new MessageHeaderCallback(new SingleResultCallback<ResponseBuffers>() {
                         @Override
-                        public void onResult(final ByteBuf result, final Throwable t) {
+                        public void onResult(final ResponseBuffers responseBuffers, final Throwable t) {
                             if (t != null) {
-                                sendFailedEvent(startTimeNanos, message, commandName, t);
+                                commandEventSender.sendFailedEvent(t);
+                                close();
                                 callback.onResult(null, t);
                                 return;
                             }
-                            ByteBufferBsonInput headerInputBuffer = new ByteBufferBsonInput(result);
-                            final MessageHeader messageHeader;
                             try {
-                                messageHeader = new MessageHeader(headerInputBuffer, description.getMaxMessageSize());
+                                boolean commandOk =
+                                        isCommandOk(new BsonBinaryReader(new ByteBufferBsonInput(responseBuffers.getBodyByteBuffer())));
+                                responseBuffers.reset();
+                                if (!commandOk) {
+                                    MongoException commandFailureException = getCommandFailureException(getResponseDocument(responseBuffers,
+                                            messageId, new BsonDocumentCodec()), description.getServerAddress());
+                                    commandEventSender.sendFailedEvent(commandFailureException);
+                                    throw commandFailureException;
+                                }
+                                commandEventSender.sendSucceededEvent(responseBuffers);
+                                T result = new ReplyMessage<T>(responseBuffers, decoder, messageId).getDocuments().get(0);
+
+                                callback.onResult(result, null);
+                            } catch (Throwable localThrowable) {
+                                callback.onResult(null, localThrowable);
```

Comment:
Weird.  It's handled by re-calling this callback, but this time with the exception that the previous invocation has thrown.  That seems confusing and it wasn't my intention.  I think it's better to handle the exception here.